plugins/org.python.pydev.parser/src/org/python/pydev/parser/grammar310/TreeBuilder310.java

```diff
@@ -648,13 +647,64 @@ public final SimpleNode onCloseNode(SimpleNode n, int arity) throws Exception {
                 Log.log("Expected arity to be > 1 here.");
                 return null;
 
+            case JJTPATTERN:
+                if (arity == 1) {
+                    SimpleNode popNode = stack.popNode();
+                    if (popNode instanceof exprType) {
+                        return new MatchValue((exprType) popNode);
+                    }
+                    return popNode;
+                } else if (arity == 2) {
+                    NameTok asname = null;
+                    patternType pattern = null;
+                    SimpleNode popNode = stack.popNode();
+                    try {
+                        asname = (NameTok) popNode;
+                    } catch (Exception e) {
+                        Log.log("Expected NameTok. Found: " + popNode);
+                    }
+                    popNode = stack.popNode();
+                    try {
+                        pattern = (patternType) popNode;
+                    } catch (Exception e) {
+                        Log.log("Expected pattern. Found: " + popNode);
+                    }
+                    return new MatchAs(pattern, asname);
+                }
+                Log.log("Expected arity to be 1 or 2 here.");
+                return null;
+
+            case JJTAS_PATTERN:
+                if (arity == 1) {
+                    Name name = null;
+                    nameTok = null;
+                    SimpleNode popNode = stack.popNode();
+                    try {
+                        name = (Name) popNode;
+                    } catch (Exception e) {
+                        Log.log("Expected name. Found: " + popNode);
```

Comment:
If something doesn't work, instead of using the plain log, there's a pattern for the parsing, which is something like:

(extracted from `makeName(int ctx)`):
```
            this.stack.getGrammar().addAndReport(
                    new ParseException("Syntax error. Expected Name, found: " + popNode.getClass().getName(), popNode),
                    "Treated class cast exception making name");
            popNode = new Name("invalid", ctx, false);
```

i.e.: it tries to fix things to be valid in the ast and reports it.

Reply:
Note: the same is valid for other places with `Log.log`.